### PR TITLE
[JAVA_API] Reorganize project structure and automatically load native libraries

### PR DESCRIPTION
**Pull request by rajatkrishna**
## Details

- *Restructure Java API project to improve code organization*
- *Move the logic for finding and loading native libraries from the **Core** class to the **NativeLibrary** class*
- *Automatically load native libraries when the class is loaded using static initializer blocks*

### CODE REVIEW

**Comment by likholat** [comment]
@rajatkrishna as I understand, we need this restructuration to easy build and use Java API with IntelliJ IDEA.
Could you also check the samples and add instructions to the Readme file on how to build and run the samples with IntelliJ IDEA?

File: modules/java_api/src/main/java/org/intel/openvino/Any.java [hunk 1]
```diff
@@ -6,6 +6,14 @@
 /** This class represents an object to work with different types. */
 public class Any extends Wrapper {
 
+    static {
```
Comment (likholat) [inline modules/java_api/src/main/java/org/intel/openvino/Any.java:9]: Too much code duplication.
As I see it, we can keep this static block only for the Wrapper superclass.

File: modules/java_api/src/test/java/org/intel/openvino/ModelTests.java
Comment (likholat) [inline modules/java_api/src/test/java/org/intel/openvino/ModelTests.java:1]: Need to review and fix samples as well.

File: modules/java_api/src/test/java/org/intel/openvino/README.md
Comment (likholat) [inline modules/java_api/src/test/java/org/intel/openvino/README.md:1]: Please add instructions on how to run tests with IntelliJ IDEA

File: modules/java_api/src/main/cpp/output.cpp [hunk 2]
```diff
@@ -1,43 +1,43 @@
...
-JNIEXPORT void JNICALL Java_org_intel_openvino_Output_delete(JNIEnv *, jobject, jlong addr)
-{
-    Output<Node> *obj = (Output<Node> *)addr;
-    delete obj;
-}
+// Copyright (C) 2020-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <jni.h> // JNI header provided by JDK
+#include "openvino/openvino.hpp"
```
Comment (dkurt) [inline modules/java_api/src/main/cpp/output.cpp:5]: That changes look like line ending format made by your IDE. Please revert the lines which did not actually change

File: modules/java_api/src/test/java/org/intel/openvino/compatibility/TestsSuite.java [hunk 3]
```diff
@@ -1,8 +1,7 @@
-package tests.compatibility;
+package org.intel.openvino.compatibility;
```
Comment (dkurt) [inline modules/java_api/src/test/java/org/intel/openvino/compatibility/TestsSuite.java:1]: Why tests share main package? I hope they are not included in the final JAR?

**Comment by dkurt** [comment]
Native libs loading work fine on Linux and Windows, but not on Mac. Can you please take a look?

@ilya-lavrenov, is that possible to run OpenVINO Mac worker on this PR?

https://github.com/dkurt/openvino_java/actions/runs/5219466114/jobs/9421327346

File: modules/java_api/src/test/java/org/intel/openvino/compatibility/TestsSuite.java (hunk 3 above)
Comment (dkurt) [reply modules/java_api/src/test/java/org/intel/openvino/compatibility/TestsSuite.java:1]: Sorry, let me rephrase, I hope that tests are not a part of package and cannot be imported in the user's app.

Comment (rajatkrishna) [reply modules/java_api/src/test/java/org/intel/openvino/compatibility/TestsSuite.java:1]: The test classes share the same package as the classes they test to provide access to package-scope classes during testing. However, to keep the tests separate from the main package during compilation and assembly, they are stored in a different physical directory- `src/test/java/` than the main package `src/main/java`. The Gradle java plugin compiles and assembles only the main package in the final JAR.

File: modules/java_api/src/main/java/org/intel/openvino/Any.java (hunk 1 above)
Comment (rajatkrishna) [reply modules/java_api/src/main/java/org/intel/openvino/Any.java]: Resolved. Thank you.

File: modules/java_api/src/main/java/org/intel/openvino/compatibility/IEWrapper.java
[the remark this answers is not in the document]
Comment (rajatkrishna) [reply modules/java_api/src/main/java/org/intel/openvino/compatibility/IEWrapper.java:1]: Resolved. Thank you.

File: modules/java_api/src/test/java/org/intel/openvino/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+## Build
+
+1. Download test data:
+```bash
+git clone https://github.com/openvinotoolkit/testdata
+```
+
+2. Create an folder `native` with native libraries in `java_api` folder. Then run `gradle` to build a package and run tests:
```
Comment (dkurt) [inline modules/java_api/src/test/java/org/intel/openvino/README.md:8]: That's done by Gradle automatically (when setupvars.sh is called).

File: modules/java_api/src/test/java/org/intel/openvino/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
+## Build
```
Comment (dkurt) [inline modules/java_api/src/test/java/org/intel/openvino/README.md:1]: Why not keep README at `modules/java_api/src/test`?

File: modules/java_api/src/test/java/org/intel/openvino/README.md (hunk 4 above)
Comment (likholat) [inline modules/java_api/src/test/java/org/intel/openvino/README.md:8]:
```suggestion
2. Run `gradle` to build a package and run tests:
```

Comment (likholat) [reply modules/java_api/src/test/java/org/intel/openvino/README.md:8]: I think we just forgot to change these instructions

File: modules/java_api/src/test/java/org/intel/openvino/README.md (hunk 5 above)
Comment (rajatkrishna) [reply modules/java_api/src/test/java/org/intel/openvino/README.md]: Moved. Thank you

File: modules/java_api/src/test/java/org/intel/openvino/CoreTests.java [hunk 6]
```diff
@@ -1,8 +1,7 @@
-package tests;
+package org.intel.openvino;
```
Comment (dkurt) [inline modules/java_api/src/test/java/org/intel/openvino/CoreTests.java:1]: @rajatkrishna, This breaks Android build: https://github.com/dkurt/openvino_java/actions/runs/5219539621/jobs/9441828252?pr=22

steps to reproduce: https://github.com/dkurt/openvino_java/blob/test_java_refactoring/.github/workflows/android.yaml

```
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/TensorTests.java:3: error: package org.junit does not exist
import static org.junit.Assert.*;
                       ^
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/TensorTests.java:5: error: package org.junit does not exist
import org.junit.Test;
                ^
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/OVTest.java:3: error: package org.junit does not exist
import org.junit.Ignore;
                ^
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/OVTest.java:4: error: package org.junit does not exist
import org.junit.Rule;
                ^
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/OVTest.java:5: error: package org.junit.rules does not exist
import org.junit.rules.TestWatcher;
                      ^
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/OVTest.java:6: error: package org.junit.runner does not exist
import org.junit.runner.Description;
                       ^
/home/runner/work/openvino_java/openvino_java/openvino_contrib/modules/java_api/src/test/java/org/intel/openvino/OVTest.java:10: error: cannot find symbol
@Ignore
 ^
  symbol: class Ignore
```

File: modules/java_api/src/test/java/org/intel/openvino/ModelTests.java
Comment (rajatkrishna) [reply modules/java_api/src/test/java/org/intel/openvino/ModelTests.java:1]: Added instructions for building and running benchmarking app, and face detection samples.

File: modules/java_api/src/test/java/org/intel/openvino/CoreTests.java (hunk 6 above)
Comment (rajatkrishna) [reply modules/java_api/src/test/java/org/intel/openvino/CoreTests.java:1]: It appears that the compilation errors are occurring due to missing JUnit dependencies. Due to these changes, gradle detects the java tests in the `src/test/java` folder and attempts to compile them. 

One solution would be to disable the compilation and execution of tests through gradle using the `-x test` flag when building for Android: https://github.com/dkurt/openvino_java/pull/24

Please let me know if you have questions or if there are alternative solutions you'd like to explore. Thank you

File: modules/java_api/src/main/cpp/output.cpp (hunk 2 above)
Comment (rajatkrishna) [reply modules/java_api/src/main/cpp/output.cpp]: Reverted. Thank you

**Comment by ravi9** [comment]
Thanks, Rajat for all the contributions. LGTM.

**Review by dkurt — APPROVED** [review]

**Review by likholat — APPROVED** [review]
